### ISSUE TITLE
LSIF and/or SCIP as output format for clangd-indexer

### ISSUE THREAD

**Issue by i-ky** [clangd/clangd#1340]
Currently `clangd-indexer` uses its own index format. But there are formats, namely [Language Server Index Format](https://microsoft.github.io/language-server-protocol/overviews/lsif/overview/) and its successor [SCIP Code Intelligence Protocol](https://github.com/sourcegraph/scip), aiming to be language- and tool-agnostic. These formats are used by tools such as GitLab's [Code Intelligence](https://docs.gitlab.com/ee/user/project/code_intelligence.html) and [Sourcegraph](https://about.sourcegraph.com/). Unfortunately, there is [only one implementation for C++/C/Obj-C](https://lsif.dev/) and it seems to be incomplete and outdated. It is based on clangd 11, so latest C++ standards are not supported. It is producing LSIF 0.4.3 while 0.6.0 is current, so resulting index may not work well with more up-to-date tools such as [VS Code LSIF extension](https://github.com/microsoft/vscode-lsif-extension).

I wonder if clangd could become a state of the art LSIF/SCIP indexer for C/C++. It has already been discussed in https://github.com/clangd/clangd/issues/587#issuecomment-785483755. https://github.com/clangd/clangd/issues/587#issuecomment-786861677 says that LSIF provides no benefit within clangd ecosystem compared to current index format and https://github.com/clangd/clangd/issues/587#issuecomment-787072306 points out that LSIF is quite bulky. Both true, but interoperability with other tools was not discussed there, so I decided to raise the question once again.

**Comment by HighCommander4** [comment]
I think clangd having the ability to emit LSIF or SCIP for a project would be valuable, as it would allow clangd to become the backend of tools that want to consume these formats.

As a concrete example, the [Searchfox](searchfox.org/) code search tool currently uses a custom clang plugin to index a C++ codebase, and then stores the results in a custom index format. This clang plugin performs many of the same functions as clangd (for example, associating declarations across different translation units, or performing heuristics to resolve dependent names inside templates). This duplication of logic could be avoided if Searchfox used clangd as its backend, with a standard interchange format like LSIF/SCIP as the interface between them.

---

This is different from having LSIF/SCIP be the format of the index clangd uses to serve interactive queries; that would require additional motivation

**Comment by i-ky** [comment]
> As a concrete example, the [Searchfox](searchfox.org/) code search tool...

GitHub interpreted this link in a weird way. I guess you meant to link to https://searchfox.org/.

**Comment by HighCommander4** [comment]
> GitHub interpreted this link in a weird way. I guess you meant to link to https://searchfox.org/.

Fixed, thanks

**Comment by i-ky** [comment]
> This is different from having LSIF/SCIP be the format of the index clangd uses to serve interactive queries; that would require additional motivation.

Thank you for clarifying it. I guess I didn't make it clear enough that the main idea is to _output_ LSIF/SCIP. I don't expect clangd to use LSIF/SCIP for its own index, I don't expect clangd to serve requests based on LSIF/SCIP generated by other tools.

**Comment by sam-mccall** [comment]
In principle this seems OK, as long as the LSIF-specific bits can be isolated and it doesn't need too many scary architectural changes.

clangd's design isn't *ideal* for this task (e.g. we've baked Preamble into too many places, you don't really need it here), but it works OK. We've used a (nonpublic) clangd-based indexer for code review workflows successfully.

However some caveats:
 - LSIF/SCIP haven't completely taken the world by storm (is there a list of applications somewhere?), which means lower impact. This could be a chicken/egg problem, but indexers *are* available, including one for C++.
 - it may be hard to achieve growth given that (unlike LSP) it's not something individual developers can adopt, and by far the largest hosting platform (github) doesn't seem inclined to adopt it. 
 - someone needs to be willing and able to build this thing. It won't be a spare-time project for me, and I think it's unlikely a priority for my employer.

If someone wants to try this, I suggest prototyping it in a fork to see roughly what we're looking at.

**Comment by tjdevries** [comment]
o/ hi everyone :grin: thanks for the ping.

(Obviously not committing to any work here) but it's definitely interesting to think about better supporting SCIP via clangd. We actually already have an (older) but somewhat interesting fork that emits LSIF (and it's on our todo list to change that to SCIP).

I'll tag in @varungandhi-src and @olafurpg since I'm currently hanging out with my brand-new baby girl on paternity leave :heart_eyes: but will be happy to pick up the conversation when I'm back in office.

**Comment by varungandhi-src** [comment]
Hey folks! 👋🏽 Varun from Sourcegraph here. I'm going to be exploring support for C++ indexing for Sourcegraph over the coming months along with other folks on my team. I'm going to summarize the state of C++ indexing as I know it to make sure we're all on the same page before discussing further:

Indexer | Base | Index format | Feature completeness
--- | --- | --- | ---
clangd | Upstream LLVM | Custom | I don't know this one. Are there known gaps for Objective-C, Windows etc.? Known perf issues? Is it robust in face of unimplemented libc++/Clang features? (I can take a look at the issue tracker here...).
[Kythe's C++ indexer](https://github.com/kythe/kythe/tree/master/kythe/cxx/indexer) | Upstream LLVM | Kythe (Protobuf) | Based on some exploration of the Chromium codebase, many common things do work, but Find References in certain headers doesn't seem to work reliably (as one example). Should support cross-language code nav with Protobuf, but it didn't seem to work for me.
[Glean's C++ indexer](https://github.com/facebookincubator/Glean/tree/main/glean/lang/clang) | libclang 12† | Glean (Thrift) | Used in prod at Meta, but I don't have any live code examples, and they don't seem to have many snapshot tests, so it is hard to tell.
[libIndexStore](https://github.com/apple/llvm-project/tree/apple/main/clang/tools/IndexStore)+[IndexStoreDB](https://github.com/apple/indexstore-db) | apple/llvm-project | Custom, shared with Swiftc | This powers navigation for Clang-based languages in Xcode; largely works well. In conjunction with Swiftc, supports cross-language code nav for Objective-C and Swift.
mozsearch/searchfox | libclang v?? | Custom (JSON) | Seems to work well based on a brief look
lsif-clang | Forked LLVM 12 | LSIF (JSON) | There are some known issues with stability. Having forked LLVM is also not great for maintenance.

†I think Meta internally uses closer to upstream LLVM/Clang for their indexer, but any related patches have not been made part of the OSS Glean distribution yet.

There are many axes to consider, in terms of indexer robustness, ease of merging changes, feature support, testing infrastructure, Bazel support/integration (e.g. distributed indexing, potentially with caching), cross-language support with Swift etc. Right now, my thinking is that we (Sourcegraph) want to prioritize robustness over other dimensions, because many Sourcegraph customers have very large amounts of code, and restarting from scratch or isolating/denylisting problematic files is impractical, especially given that there is a not-too-high soft upper bound on indexing speed due to Clang's semantic analysis perf.

I've been thinking of having a multi-process architecture so that any crashes when indexing individual files do not affect the rest of the indexing process. Right now, clangd uses multi-threading (and so does lsif-clang); IIUC it'd be hard to provide the same level of robustness with multi-threading (but maybe clangd is robust enough even for very large codebases that we can get away with this?). Such a design would run counter to this requirement:

> In principle this seems OK, as long as the LSIF-specific bits can be isolated and it doesn't need too many scary architectural changes.  

Apart from that, my prior experience contributing smaller patches to upstream LLVM about 1~2 years ago has been that it can be hard to find reviewers for non-small patches if your work is not a priority for some maintainer (and maintainers tend to be understandably quite busy with their own work commitments), the buildbots are noisy and require aggressive email filtering, and it is hard to understand if a failure is due to your change or not (because most/all testing was post-merge).

One of the other potential problems with SCIP support in upstream clang(d) is third-party dependencies. AFAIK LLVM has historically been reluctant to add non-optional third-party dependencies, so this means that we'd have to maintain a separate build configuration that allows using protoc, or check in a generated header, and then only have a test dependency on protoc to make sure that the generated header is in sync with the schema. I'm not sure if clangd is considered as "core" or not, if it isn't, then maybe we don't need a separate build config... I see that openmp does have a [proto file](https://github.com/llvm/llvm-project/blob/main/openmp/libomptarget/plugins/remote/include/openmp.proto)...

All-in-all, I think clangd is one option we can consider, but there's many different options here, and it's not clear to me that building on top of clangd/adding SCIP support to clangd is necessarily the best one. This is something we'll be exploring over the coming weeks. In the interest of transparency, one of the other options we've been looking into is adding support for Kythe to SCIP conversion, since Kythe's indexer has been battle-tested at Google.

---

Want to briefly address one of the other points mentioned up thread:

> LSIF/SCIP haven't completely taken the world by storm (is there a list of applications somewhere?), which means lower impact. This could be a chicken/egg problem, but indexers are available, including one for C++.

SCIP is relatively new at this point (about ~6 months old). In the meantime, we've ported the following indexers to SCIP:

> [scip-java](https://github.com/sourcegraph/scip-java) (for Java, Kotlin and Scala), [scip-typescript](https://github.com/sourcegraph/scip-typescript), [scip-python](https://github.com/sourcegraph/scip-python), [scip-ruby](https://github.com/sourcegraph/scip-ruby) and [rust-analyzer](https://github.com/sourcegraph/rust-analyzer).

We've also been adding [docs](https://docs.sourcegraph.com/code_navigation/explanations/writing_an_indexer) & [tooling](https://github.com/sourcegraph/scip/blob/main/docs/CLI.md) to help SCIP adoption. But yeah, it is still early days. 😄

**Comment by i-ky** [comment]
> This is something we'll be exploring over the coming weeks.

@varungandhi-src, is there an issue or discussion I can subscribe to? I would like to be up-to-date on this topic.

**Comment by i-ky** [comment]
> * by far the largest hosting platform (github) doesn't seem inclined to adopt it.

It is true that [GitHub's native code navigation](https://docs.github.com/en/repositories/working-with-files/using-files/navigating-code-on-github) uses a different technology, but it [lacks C/C++ support](https://github.com/community/community/discussions/8594). At the same time [Sourcegraph provides browser extensions](https://docs.sourcegraph.com/integration/browser_extension) that replace native GitHub code navigation with LSIF/SCIP based one. So for C/C++ projects on GitHub LSIF/SCIP code navigation is the only possible option, although adopting it requires some steps on projects' and users' side.

**Comment by sam-mccall** [comment]
@tjdevries Congratulations, hope your daughter is doing well and you're all getting some sleep here and there!

@varungandhi-src thanks for laying out the things you're looking at, this is really interesting.

---

Evaluating clangd:

> Are there known gaps for Objective-C

We do use clangd with Obj-C at Google, it's a bit less mature than C++ but not bad. 

> Windows etc.? [...] Is it robust in face of unimplemented libc++/Clang features? 

I think the answer here is likely to be the same for ~all clang-based indexers: it'll be at most as robust as clang, and you need to get it just the right parse flags.

Similarly most of the invalid-code-resilience is happening in clang and will be similar. (We added some substantial portions for clangd, but it's enabled everywhere I think).

In terms of random crashes on **top** of clang, clangd may well have some as we don't systematically run it on whole codebases to stamp out bugs. OTOH it's relatively well tested on broken code. (Compare to e.g. kythe which is regularly run on huge snapshots of correct code).

> Known perf issues? (I can take a look at the issue tracker here...).

For clangd the biggest issue is that the indexing subsystem is designed for throughput, but the AST that features are served from is optimized for latency: assuming preambles are built and will be reused.

The "obvious" thing is to modify clangd's indexer to spit out SCIP, but clangd's index isn't designed to serve features from directly, but rather to support an AST. So you'll have a very small feature set. I think this is what `lsif-clang` is.

The "nice" thing is to build the AST as if opening the file in clangd, and then run queries against it. However I'd guess this isn't ideally fast: not only the unneccesary preamble (which could probably be made optional), but also e.g. "find what's under the cursor" isn't trivially cheap so running it in a loop is an inefficient way to gather hover info.
That said, we *do* use this approach in production for code review indexing.

---

I also have a bit of direct experience with Kythe:
 - as a user (of our internal browser, very similar to source.chromium.org)
 - consuming the data (e.g. we use it for xrefs in google's internal clangd)
 - indexing projects (a working prototype of indexing llvm-project with GH)

It's basically great. It's mature and works. The multi-language support is for real, and at least for the languages heavily used at Google the indexers are great. (I guess I can only personally vouch for C++, Go, Java, Protobuf, but the Python/JSTS crowd seem to be happy).  Cross-lang nav also works well for protos if set up right (which is a big deal for Google, probably less for others), I don't know how well that feature generalizes to other interop situations.

In your place this would be my choice I think. It seems well aligned with your tech stack (you mentioned Bazel and SCIP is protos).

XRefs are the biggest strength (and probably most important feature?). I'd say better than clangd, e.g. kythe can deal with codebases where the same symbol is defined several times but never linked together in the same binary. (`main()`, but also just name conflicts).

Hover is IMO relatively weak compared to clangd - it's fairly consistent across languages but I think lacking in detail (maybe because the relevant detail differs across languages?).

I'm not sure how Kythe would be for semantic highlighting, I don't think we use it for that.

The data model is a bit to get your head around, at least for me, but it's pretty rich and flexible.

---

>> LSIF/SCIP haven't completely taken the world by storm [...]
>
> SCIP is relatively new at this point (about ~6 months old). In the meantime, we've ported the following indexers to SCIP:

Sorry, I definitely didn't mean to be dismissive and don't think I expressed my point well. You're right that SCIP is new, and I think the rationale and tech differences from LSIF are well-thought-out. 

Mostly I mean tools to **produce** the data exist in many languages, but (anecdotally) relatively few tools to **consume** the data, and few people producing the data. So it may produce a big open ecosystem, but when comparing to LSP we should bear in mind that it hasn't *yet*, and previous attempts (kythe?) have had limited success.

(I hope it succeeds! Honestly I've underestimated the `sourcegraph.com` instance - if I can easily upload project indexes to it and consume them from github browsing/phab reviews it looks really useful. Unfortunately it gives the appearance of being a "free demo" rather than something to be used seriously for open-source, and I can't work out whether/how I can upload project metadata to it. The tech & UI looks solid, it's really the number of disguised links to marketing funnels, "get free trial" etc).

**Comment by asutherland** [comment]
Chiming in for Mozilla's searchfox (AKA [mozsearch](https://github.com/mozsearch/mozsearch)) to fill in gaps / provide relevant context:

#### Implementation details:
- Searchfox's [C++ indexer](https://github.com/mozsearch/mozsearch/blob/master/clang-plugin/MozsearchIndexer.cpp) gets kept up-to-date with clang trunk [as part of mozilla's build team's work to keep us compatible with clang trunk](https://bugzilla.mozilla.org/show_bug.cgi?id=1614858) with v14 currently used for production.
- Searchfox's native [analysis format](https://github.com/mozsearch/mozsearch/blob/master/docs/analysis.md) is newline-delimited JSON[1] file where each file corresponds exactly to a source/generated file and each record is associated directly with a (spelling) token in that file.  Extra semantics like structured class information is associated with the canonical definition token.  The C++ indexer uses file-locking and de-duplicates identical JSON records, which handles headers.  We run the analysis for our 4 tier1 platforms and [merge](https://github.com/mozsearch/mozsearch/blob/master/tools/src/file_format/merger.rs) the records.
  - These files get ingested into a [crossref file](https://github.com/mozsearch/mozsearch/blob/master/docs/crossref.md) whose format is roughly `SYMBOL\n{ JSON }\n` that gets binary searched for lookup although larger JSON values are replaced with offsets into a similar secondary file to improve binary search performance against the primary file.

#### Adoption of SCIP:
- Searchfox is [moving to use rust-analyzer's SCIP output for our rust indexing](https://bugzilla.mozilla.org/show_bug.cgi?id=1761287) and our hope is also [to move to ingesting SCIP from scip-typescript for JS](https://bugzilla.mozilla.org/show_bug.cgi?id=1740290) as well as ideally adopt scip-python for entirely new language support as well as potentially use scip-java over a searchfox-specific working Java indexer that has been written but not integrated.
  - Our current approach is to render the SCIP format to searchfox's custom analysis format for subsequent ingestion into the crossref database and for direct consumption via HTML file rendering.
- Searchfox is interested in in moving to clangd-produced SCIP, noting that this would probably be hybridized where we would initially leverage our existing analysis record merging mechanism to fuse clangd's output with the existing searchfox C++ indexer analysis output.
- Mozilla has a number of specialized Interface Definition Languages (IDLs) and Domain Specific Languages (DSLs) where we already have searchfox indexing support and are looking to add support for more languages, but where we would also place a high value on Language Server Protocol (LSP) integration and would be willing to overhaul the pipeline to directly produce SCIP if that provides the most pragmatic/reusable path to exposing those languages to VS Code / other IDEs.

1: Note that any [rendered file](https://searchfox.org/mozilla-central/source/dom/base/nsGlobalWindowInner.cpp) also has the [analysis available if you replace /source/ with /raw-analysis/](https://searchfox.org/mozilla-central/raw-analysis/dom/base/nsGlobalWindowInner.cpp).

**Comment by varungandhi-src** [comment]
Thanks @sam-mccall and @asutherland for chiming in with details, this is very helpful. I'll break out my responses into separate comments.

Sam:

> However I'd guess this isn't ideally fast: not only the unneccesary preamble (which could probably be made optional), but also e.g. "find what's under the cursor" isn't trivially cheap so running it in a loop is an inefficient way to gather hover info. That said, we do use this approach in production for code review indexing.

Yeah, we ran into this when implementing scip-ruby on top of Sorbet as well; reusing Sorbet's LSP code would almost certainly have given us bad perf for indexing, because what we need is 1~2 top-down traversals of the AST, and hovers are implemented in a much heavier way (which is fine for an IDE because it isn't a bulk API).

The points you mention about the preamble etc. resonate with some of the pains we encountered in the past. Earlier indexers were built on top of language servers, but that made them harder to understand, maintain and optimize for perf (I think this is at least directionally accurate, but I don't have first-hand knowledge since this was before I joined Sourcegraph).

For code review, I'm thinking we would like to lean on the build system to reuse the available incrementality and early cutoff, but so far, we haven't tackled that, our focus has been on from-scratch indexing perf and language coverage.

---

On Kythe, yeah, the points you mention were some of the reasons we started exploring it. My and @olafurpg's initial impressions weren't so great; we indexed a hello world C++ project and generated index was about 60MB IIRC (6 MB after zipping). This includes the standard library, so we have to discount that, but it still seems a bit too heavy. You mention this:

> The data model is a bit to get your head around, at least for me, but it's pretty rich and flexible.

This does seem to be a potential issue. One of the big reasons we created SCIP instead of sticking with LSIF or moving to targeting Kythe for indexers was that having a simpler data format would increase the ease of writing, maintaining and testing indexers. We found the flexibility of having graphs in the data format (especially with the pervasive use of integer IDs in LSIF) made debugging issues and on-boarding people much harder. From an intuitive perspective, having graphs makes total sense, but pragmatically, the "arrays of structs + extra relationships" approach in SCIP has been much easier to work with. The flexibility of a graph format also makes writing additional tooling on top harder. For example, we've been able to add [lint checks](https://github.com/sourcegraph/scip/blob/main/cmd/lint_test.go#L98) for SCIP much more easily than we could've done for LSIF.

One of the interesting bits about Kythe is the separation of the extraction and indexing steps, which is not how other SCIP indexers are designed today. Consuming compilation steps extracted by Kythe is also something we can explore, without reusing the indexer, as this could potentially enable Bazel integration in a more plug-and-play way across languages.

---

Could you clarify this a bit?

> indexing projects (a working prototype of indexing llvm-project with GH)

Do you mean indexing llvm-project with Kythe's C++ indexer and the GH is a typo? Or are you referring to some prototype of GitHub integration (with Kythe)?

---

> Sorry, I definitely didn't mean to be dismissive and don't think I expressed my point well

No worries! I didn't interpret your comment as being dismissive, I was only trying to provide context on when SCIP came about and the progress we've made so far within that time period. Totally understand your concerns around adding support for what is otherwise not-a-widely-adopted data format.

> Unfortunately it gives the appearance of being a "free demo" rather than something to be used seriously for open-source, and I can't work out whether/how I can upload project metadata to it

Thanks for the feedback, I think we can be more explicit about this in our docs. The TLDR is that if you [sign up](https://sourcegraph.com/sign-up) on Sourcegraph.com using your GitHub/GitLab account, then you can generate an index using any indexer and you can upload an index using the [Sourcegraph CLI](https://github.com/sourcegraph/src-cli) using `src code-intel upload -github-token=blah` (or `-gitlab-token`) for the repos you own/maintain (I forget the exact rules around OSS GitHub orgs).

**Comment by varungandhi-src** [comment]
For @asutherland points about Searchfox:

> Searchfox's [C++ indexer](https://github.com/mozsearch/mozsearch/blob/master/clang-plugin/MozsearchIndexer.cpp?rgh-link-date=2022-10-20T21%3A30%3A03Z) gets kept up-to-date with clang trunk [as part of mozilla's build team's work to keep us compatible with clang trunk](https://bugzilla.mozilla.org/show_bug.cgi?id=1614858) with v14 currently used for production.

That's good to know. I originally has the wrong impression that you were using libclang, but based on the headers being included and the fact that it is a plugin explain how searchfox is able to be precise. 😅

> Searchfox's native [analysis format](https://github.com/mozsearch/mozsearch/blob/master/docs/analysis.md?rgh-link-date=2022-10-20T21%3A30%3A03Z) is ...

Thanks for sharing the links (and writing those docs!). I spent some time reading the docs and the code, and it looks like the JSON format is in some ways a slightly flatter cousin of SCIP.

> Adoption of SCIP:

Exciting! I've subscribed to the Rust PR; looking forward to featuring searchfox in the SCIP README once that lands. :)

> Searchfox is interested in in moving to clangd-produced SCIP

Just to clarify things here, is your broad point that if there were a new SCIP indexer for C++, you would consider using it, or do you specifically prefer that it be built on top of/as part of clangd? I'm asking because some other options under consideration are using searchfox as a starting point, or starting from scratch. Building on top of clangd seems possible, but as Sam has pointed out, it doesn't seem like the most straightforward thing because clangd has somewhat different requirements. OTOH, searchfox seems arguably better suited as a foundation because it is an indexer, and the output format seems to share a few similarities with SCIP, instead of targeting a graph format (unlike Kythe).

**Comment by khuey** [comment]
> Mostly I mean tools to **produce** the data exist in many languages, but (anecdotally) relatively few tools to **consume** the data, and few people producing the data. So it may produce a big open ecosystem, but when comparing to LSP we should bear in mind that it hasn't _yet_, and previous attempts (kythe?) have had limited success.

I maintain https://pernos.co/ which is capable of ingesting LSIF to provide a more IDE-like experience in the debugger (e.g. Go to Definition, etc).

The biggest barrier to adoption for us is the lack of good *producer* tools. Generating LSIF for C++ (which is most of our customer base) is only possible afaik by installing Sourcegraph's forked toolchain which (and correct me if I'm wrong) is several LLVM versions out of date at this point.

From reading Sourcegraph's announcement blog post SCIP seems to be an improvement largely for producers and not consumers. Performance isn't a big concern for us at this point (consuming the LSIF is roughly two orders of magnitude faster than rust-analyzer is at generating it). That said it seems like it would not be much work to also consume SCIP (or, alternatively, to just feed SCIP through the SCIP->LSIF converter) so either output format seems fine for our needs.

**Comment by sam-mccall** [comment]
Pernosco looks interesting, it's exactly the sort of tool that I wasn't aware existed, and it's clear why a C++ indexer in particular is needed. Thanks!

>  Earlier indexers were built on top of language servers, but that made them harder to understand, maintain and optimize for perf

Yeah, bolting it on top of a language server seems like an ugly compromise in the end. There are definitely sharable pieces (traversals like highlighting/inlay hints, local operations like "find hover text for this decl", infra pieces like "iterate over the references in this AST", heuristic resolution of dependent code).

So if you think there's enough value in clangd's versions of these things, either trying to build a SCIP indexer within the clangd codebase (refactoring bits as needed), or lifting components up into clang/Tooling or so, sounds reasonable to me.

Of course if most of it is new code, it might make a lot more sense to just start fresh, or use searchfox as a base, or whatever. 

> [Kythe] My and @olafurpg's initial impressions weren't so great; we indexed a hello world C++ project and generated index was about 60MB IIRC (6 MB after zipping). This includes the standard library, so we have to discount that, but it still seems a bit too heavy

I guess just make sure you're comparing apples to apples here. The C++ standard library + the c headers it depends on is huge (10MB for just libstdc++ 12/* and 12/bits/* on my system), if you were indexing *all* of it I think 60MB sounds very reasonable. That said, I can totally believe that the data format is redundant & verbose. But it's just an intermediate, right?

> Could you clarify this a bit?
>
> > indexing projects (a working prototype of indexing llvm-project with GH)

Sorry, this was supposed to be "GH actions" - just running the Kythe indexer to generate an artifact we could feed into code search. Not a big deal, but the sort of thing that needs to work reliably.

**Comment by asutherland** [comment]
Replying to @varungandhi-src 
> > Searchfox is interested in in moving to clangd-produced SCIP
> 
> Just to clarify things here, is your broad point that if there were a new SCIP indexer for C++, you would consider using it, or do you specifically prefer that it be built on top of/as part of clangd? I'm asking because some other options under consideration are using searchfox as a starting point, or starting from scratch. Building on top of clangd seems possible, but as Sam has pointed out, it doesn't seem like the most straightforward thing because clangd has somewhat different requirements. OTOH, searchfox seems arguably better suited as a foundation because it is an indexer, and the output format seems to share a few similarities with SCIP, instead of targeting a graph format (unlike Kythe).

The core motivating concerns behind this statement are organizational in nature:
- Our users are using searchfox and VS code/language-server-consuming IDEs and we have explicit support for [setting up a checkout by running `mach ide vscode`](https://firefox-source-docs.mozilla.org/contributing/vscode.html) which configures clangd.  There are clear benefits to maximizing consistency of cross-referencing between these two experiences.
- Engineering management would like to maximize the synergy of any internal investments we make between the two.  In particular, coordination around enhancements to searchfox's C++ indexer template handling have raised questions in this area.  Hand-wavingly, it feels like there would be more interest in allocating engineers to contribute to clangd.  The existence of VS code on the web via github.dev has shaped discussions in a complicated fashion as I believe the use-cases are disparate ([see my searchfox vision blog post](https://www.visophyte.org/blog/2022/10/05/andrews-searchfox-vision-2022/)) but our resources to allocate to this are limited.

I definitely understand that the differing requirements mean that indexers like searchfox and sourcegraph potentially are working against the grain of the existing clangd requirements.  I'd personally be open to migrating from searchfox's custom clang plugin indexer to a SCIP based one even if it was not maximally clangd, but would likely only do this when there was a clear benefit to making the change.  Looking at our current indexer run times I see that Firefox x64 debug builds without indexing are currently taking 12 minutes but our searchfox indexer builds are taking 54 minutes, suggesting our file locking approach (at least with our current set of includes) is not scaling well and it would be an easy win to move to a SCIP indexer that was more clever like having a daemon that the clang plugin could stream its output at which would be responsible for de-duplicating records and/or pre-emptively suppressing traversals.  I should note that we are also doing a lot of rust save-analysis generation at the same time on the same VM and I don't know how that impacts overall performance.  I'm happy to discuss this more on the sourcegraph discord or at https://github.com/sourcegraph/sourcegraph/issues/42280 as this specific thread is probably veering too off-topic for this repo.

**Comment by varungandhi-src** [comment]
@khuey: Thanks for mentioning Pernosco; I think I'd seen one of your tweets about LSIF but forgot about it.

> The biggest barrier to adoption for us is the lack of good producer tools. Generating LSIF for C++ (which is most of our customer base) is only possible afaik by installing Sourcegraph's forked toolchain which (and correct me if I'm wrong) is several LLVM versions out of date at this point.

Yes, lsif-clang is on LLVM 12. We're looking to fix that gap.

> From reading Sourcegraph's announcement blog post SCIP seems to be an improvement largely for producers and not consumers.

This is mostly correct, since the number of producers so far is much more than the number of consumers. However, we are [going to ingest SCIP natively in our backend](https://github.com/sourcegraph/sourcegraph/issues/42183), and our [frontend also consumes SCIP](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/lsif/html.ts?L3:89) for syntax highlighting certain languages. So we do have some interest in it being easy to consume too (and if you run into issues as a consumer, you're welcome to file issues 😄).

---

Responding to @sam-mccall:

> I guess just make sure you're comparing apples to apples here. The C++ standard library + the c headers it depends on is huge (10MB for just libstdc++ 12/* and 12/bits/* on my system), if you were indexing all of it I think 60MB sounds very reasonable. That said, I can totally believe that the data format is redundant & verbose. But it's just an intermediate, right?

Thanks for flagging this; I was underestimating the size of the C++ standard library. With that in perspective, it doesn't seem so bad. I should've clarified this bit earlier; our broader concern wasn't around the raw size itself (it does compress well), but rather (a) tooling/especially diffs (right now, we can get pretty decent diffs on SCIP indexes -- although, we usually look at annotated source code snapshots in tests, diffing indexes is sometimes important) (b) on-boarding & complexity (SCIP having less structure makes understanding it easier, and we have limited engineering bandwidth). The large-feeling raw size kinda' strengthened that impression ("wow, there's a lot going on here"), but maybe it's not so bad. I will take a second look.

---

Responding to @asutherland:

Thanks for the context. The bits about editor integration and not wanting bifurcated tooling make a lot of sense. I'm always glad when people mention the downsides/drawback of some software up-front, so I'm glad you brought up the perf issue. For all our other major indexers, indexing perf is typically within 1x~2x of type-checking speed, so that would definitely be a problem for us as well.

---

Wrapping up, I'm will try to report back here after a month of two on what we end up settling on for C++ indexing, after trying out a few different things. You can also subscribe to the [Investigate approaches for a SCIP indexer for C++ issue](https://github.com/sourcegraph/sourcegraph/issues/42280) in case you'd like more regular updates (maybe around weekly or so, assuming I'm not pulled into something else during the week).

**Comment by khuey** [comment]
> So we do have some interest in it being easy to consume too (and if you run into issues as a consumer, you're welcome to file issues smile).

Just to be clear here my point was "as a consumer I don't think I care whether we get the data in the LSIF or SCIP format as long as we get it" not "SCIP is worse for us as a consumer".

**Comment by i-ky** [comment]
> "find what's under the cursor" isn't trivially cheap so running it in a loop is an inefficient way to gather hover info.
> That said, we _do_ use this approach in production for code review indexing.

I guess this kind of approach will be used by https://github.com/microsoft/lsif-node/issues/134 if it will be implemented.

**Comment by kadircet** [comment]
hey folks, as discussed throughout the issue, we're definitely supportive of such a tool but it's not in our priority list to get to it soon and it'll require a significant amount of work to get it going. so if someone decides to follow up on the ideas here and turn it into implementation feel free to re-open.
